ballerina/tests/15-xa-transaction.bal

```diff
@@ -0,0 +1,159 @@
+    }
+
+    int count1 = check getCustomerCount(dbClient1, 1);
+    int count2 = check getSalaryCount(dbClient2, 1);
+    test:assertEquals(count1, 1, "First transaction failed");
+    test:assertEquals(count2, 1, "Second transaction failed");
+
+    check dbClient1.close();
+    check dbClient2.close();
+}
+
+@test:Config {
+    groups: ["transaction", "xa-transaction"]
+}
+function testXATransactionFailureWithDataSource() returns error? {
+    Client dbClient1 = check new (HOST, USER, PASSWORD, DATABASE, PORT, connectionPool = {maxOpenConnections: 1},
+                                    options = {
+                                        useXADatasource: true
+                                    }
+                                   );
+    Client dbClient2 = check new (HOST, USER, PASSWORD, DATABASE, DBCLIENT2_PORT, connectionPool = {maxOpenConnections: 1},
+                                    options = {
+                                        useXADatasource: true
+                                    }
+                                   );
+
+    transaction {
+        // Intentionally fail first statement
+        _ = check dbClient1->execute(`insert into CustomersTrx (customerId, name, creditLimit, country)
+                                values (30, 'Anne', 1000, 'UK')`);
+        _ = check dbClient2->execute(`insert into Salary (id, value ) values (10, 1000)`);
+        check commit;
+    } on fail error e {
+        test:assertTrue(e.message().includes("Duplicate"), msg = "Transaction failed as expected");
+    }
+
+    int count1 = check getCustomerTrxCount(dbClient1, 30);
+    int count2 = check getSalaryCount(dbClient2, 20);
+    test:assertEquals(count1, 1, "First transaction should have failed");
+    test:assertEquals(count2, 0, "Second transaction should not have been executed");
+
+    check dbClient1.close();
+    check dbClient2.close();
+}
+
+@test:Config {
+    groups: ["transaction", "xa-transaction"]
+}
+function testXATransactionPartialSuccessWithDataSource() returns error? {
+    Client dbClient1 = check new (HOST, USER, PASSWORD, DATABASE, PORT, connectionPool = {maxOpenConnections: 1},
+                                    options = {
+                                        useXADatasource: true
+                                    }
+                                   );
+    Client dbClient2 = check new (HOST, USER, PASSWORD, DATABASE, DBCLIENT2_PORT, connectionPool = {maxOpenConnections: 1},
+                                    options = {
+                                        useXADatasource: true
+                                    }
+                                   );
+
+    transaction {
+        _ = check dbClient1->execute(`insert into Customers (customerId, name, creditLimit, country)
+                                values (30, 'Anne', 1000, 'UK')`);
+        // Intentionally fail second statement
+        _ = check dbClient2->execute(`insert into SalaryTrx (id, value ) values (20, 1000)`);
```

Comment:
```suggestion
        _ = check dbClient2->execute(`insert into SalaryTrx (id, value) values (20, 1000)`);
```

Reply:
Fixed in [f03e97e](https://github.com/ballerina-platform/module-ballerinax-oracledb/pull/466/commits/f03e97e3552caf32690516dbe0aed1b50c60bea0)